docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+* Power: We can parse the CEL expression and examine the AST ourselves to extract useful information such as 
+  which secrets they're attempting to use, which enables us to deliver on the "no secret APIs called if no secrets used"
+  goal.
+
+**Cons**
+
+* May not be as familiar to users.
+* No string interpolation, which makes the templates not quite as clean as they otherwise could be.
+  * Mitigation: We could write our own, although then that function will _definitely_ not be familiar to users. 
+
+#### Option 3: Write our own simple text-replacement engine
+
+Supporting a format similar to `text/template`, but without any of the other looping/flow-control features.
+`DefaultEndpointsProtocol=https;AccountName={{.Spec.AzureName}};AccountKey={{.Spec.OperatorSpec.Secrets.Key1}};EndpointSuffix=core.windows.net`
+
+**Pros**
+
+* Theoretically simple
+
+**Cons**
+
+* More work than using something off the shelf. Including things like getting errors right, etc.
+* Not as easily extensible as something like CEL if we wanted to add more capabilities.
+* Trusting our security/hardening is scarier than trusting a "real" product.
+
+### Location
+
+#### Option 1: Part of operatorSpec.secrets and operatorSpec.configMap
+
+A new `formattedValues` option in `operatorSpec.secrets` and `operatorSpec.configMaps`
+
+**Secrets**
+```yaml
+  operatorSpec:
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+      formattedValues:
+        - name: redis-secret # Name of the destination secret
+          key: hostPort # Name of the key in the secret
+          value: "'%s:%s'.format([secret.hostName, secret.sslPort])" # Value (format) of the secret
+```
+
+**ConfigMaps**
+```yaml
+  operatorSpec:
+    configMaps:
+      formattedValues:
+        - name: account-data # Name of the destination configMap
+          key: id # Name of the key in the configMap
+          value: "resource.status.id" # Value (format) of the configMap entry
+```
+
+**Pros**
+
+* Fits into our existing model relatively well.
+
+**Cons**
+
+* Need to deal with possibility of property name collisions with `formattedValues`,
+  though practically it seems unlikely to be common.
+
+#### Option 2: Two new operatorSpec properties: operatorSpec.formattedSecrets and operatorSpec.formattedConfigMaps
+
+Other than the location, this is very similar to the
+[first proposal](#option-1-part-of-operatorspecsecrets-and-operatorspecconfigmap).
+
+**Secrets**
+```yaml
+  operatorSpec:
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+    formattedSecrets:
+      values:
+        - name: redis-secret # Name of the destination secret
+          key: hostPort # Name of the key in the secret
+          value: "'%s:%s'.format([secret.hostName, secret.sslPort])" # Value (format) of the secret
+```
+
+**ConfigMaps**
```

Comment:
Would we be able to export collections? Most preferably `Maps` since we don't support exporting slices yet.

Reply:
Yes, though the user might need to write a CEL expression to explain what exactly they want to gather from the map

Reply:
I will update the documentation here to be explicit about how this might look in an actual CEL expression

Reply:
Updated the FAQ documentation. It now says:

> Q: Does this work with arbitrary arrays and maps?
>
> A: Yes! The syntax for maps is: `request.myMap['hello']`, and for arrays `request.myArray[0]`. CEL has a concept of
macros and there are some useful ones such as `filter`, `exists`, and `all` 
[built in](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros), so you can craft more complex
expressions like `request.slice.exists_one(i, i.name=='foo') ? request.slice.filter(i, i.name == 'foo')[0].value : 0`
which checks if the slice contains exactly one element whose name is `foo` and if so, gets that elements value, 
otherwise returns 0. A similar construct can be done for maps using the same `exists_one` and `filter` macros.

Reply:
Ah, but what about if you want the _whole_ collection...

Reply:
Ok we can support whole collections too. Here's how:
> Q: Can this be used to export entire collections of secret or configmap values?
> 
> A: Yes. We will support two flavors of entry into the `dynamicSecrets` and `dynamicConfigMaps` collections. One
flavor is the one we've seen above, where `value` is a CEL expression that returns a `string` and the `name` and `key`
fields specify which configMap or secret and what key within that to export.
>
> The other kind of entry is:
> ```
>  operatorSpec:
>    dynamicConfigMaps:
>      - name: account-data # Name of the destination configMap
>        valueMap: "{"key1": "foo", "key2": "bar"}" # CEL expression that returns a map[string]string
> ```
>
>The `valueMap` fields result can be directly saved as keys + values in the resulting configmap or secret, so the author
of the CEL expression has direct control over what values go into their secret/configMap.
>


Great question @super-harsh.